src/main/scala/scalismo/faces/segmentation/LoopyBPSegmentation.scala

```diff
@@ -0,0 +1,493 @@
+    }
+
+    // multiply all messages (leaf one out is done by dividing later)
+    val allRegionProduct: LogRegionTypeDist = incomingRegionPriorMessages.toUnsafeImage.values.map(_.toLogLabelDist).reduce(_ * _).normalized
+
+
+    outgoingRegionPriorMessages.transformWithIndexParallel { (x, y, _) =>
+      val incomingMessages: LogRegionTypeDist = allRegionProduct / incomingRegionPriorMessages(x, y).toLogLabelDist
+      LabelDist.tabulate(numLabels) { (z: Label) =>
+        incomingMessages.marginalize((c: Label) => regionPrior(c.value)(x, y)(z))
+      }.normalized
+    }
+  }
+
+  /** a single message pass in given direction */
+  private def messagePass(messageField: MessageFieldBuf, localMessages: PixelImage[LabelDist], binaryDist: BinaryLabelDistribution, numLabels: Int, direction: Direction): Unit = {
+    direction match {
+      case Right => // do rows in parallel, sequential along row
+        for {row <- (0 until messageField.height).par
+             x <- 0 until messageField.width - 1} {
+          // store message: is incoming from the left in next node
+          messageField(x + 1, row)(Left.toInt) = calculateMessage(messageField, localMessages, binaryDist, numLabels, x, row, Right)
+        }
+      case Left => // do rows in parallel, sequential along row
+        for {row <- (0 until messageField.height).par
+             x <- messageField.width - 1 to 1 by -1} {
+          messageField(x - 1, row)(Right.toInt) = calculateMessage(messageField, localMessages, binaryDist, numLabels, x, row, Left)
+        }
+      case Down => // do columns in parallel, sequential along column
+        for {col <- (0 until messageField.width).par
+             y <- 0 until messageField.height - 1} {
+          messageField(col, y + 1)(Up.toInt) = calculateMessage(messageField, localMessages, binaryDist, numLabels, col, y, Down)
+        }
+      case Up => // do columns in parallel, sequential along column
+        for {col <- (0 until messageField.width).par
+             y <- messageField.height - 1 to 1 by -1} {
+          messageField(col, y - 1)(Down.toInt) = calculateMessage(messageField, localMessages, binaryDist, numLabels, col, y, Up)
+        }
+    }
+  }
+
+  def estimateColorDists(image: PixelImage[RGB], labelDist: PixelImage[LabelDist]): IndexedSeq[ColorDistribution] = {
+    // max only estimation currently
+    val maxLabels = labelDist.map {
+      _.maxLabel
+    }
+    val maxLabel: Int = labelDist.map {
+      _.length
+    }.values.max
+    // estimate for each label
+    IndexedSeq.tabulate(maxLabel) { n =>
+      val colors = image.zip(maxLabels).values.collect { case (color, label) if label == n => color }.toIndexedSeq
+      if (colors.nonEmpty)
+        GaussDist(colors)
+      else
+        UniformDist
+    }
+  }
+
+  def main(args: Array[String]): Unit = {
+
+    implicit val rnd = Random(37)
+
+    val image = PixelImageIO.read[RGB](new File("/home/eggebe01/facedb/sandro.png")).get.resample(256, 256)
```

Comment:
remove path not accessible from the repo.

Reply:
Would you prefer to add a testImage in the test/ressources of the repo, or remove the main completely?

Reply:
Without having looked at the code closely i think this is not the right place for an application. So I propose to either make it a working test for the code using an image where you know what should come out so you can compare the result or remove the main.

Reply:
main removed